// grunt-vtex-deploy/Gruntfile.js
/*
 * grunt-vtex-deploy
 * https://github.com/guilherme/grunt-vtex-deploy
 *
 * Copyright (c) 2013 Guilherme Rodrigues
 * Licensed under the MIT license.
 */

'use strict';

module.exports = function(grunt) {

  // Project configuration.
  grunt.initConfig({

    // Before generating any new files, remove any previously-created files.
    clean: {
      tests: ['tmp']
    },

    coffee: {
      main: {
        files: {
          'tasks/vtex_deploy.js':'tasks/vtex_deploy.coffee'
        }
      }
    },

    // Configuration to be run (and then tested).
    vtex_deploy: {
      main: {
        options: {
          indexPath: 'build/index.html',
          whoamiPath: 'whoami',
          includeHostname: {
            hostname: 'io.vtex.com.br',
            files: ["build/index.html", "build/test1"]
          }
        }
      },
      walmart: {
        options: {
          buildDirectory: 'build-raw',
          bucket: 'vtex-io-walmart',
          requireEnvironmentType: 'stable',
          indexPath: 'build-raw/index.html',
          indexOnRoot: true,
          includeHostname: {
            hostname: 'VTEX_IO_HOST',
            files: ["build-raw/index.html", "build-raw/test1"]
          }
        }
      }
    },

    // Unit tests.
    nodeunit: {
      tests: ['test/*_test.js']
    }

  });

  // Actually load this plugin's task(s).
  grunt.loadTasks('tasks');

  // These plugins provide necessary tasks.
  grunt.loadNpmTasks('grunt-contrib-clean');
  grunt.loadNpmTasks('grunt-contrib-nodeunit');
  grunt.loadNpmTasks('grunt-contrib-coffee');

  // Whenever the "test" task is run, first clean the "tmp" dir, then run this
  // plugin's task(s), then test the result.
  grunt.registerTask('test', ['clean', 'vtex_deploy']);

  // By default, lint and run all tests.
  grunt.registerTask('default', ['test']);

};
